library/core/src/mem/maybe_uninit.rs

```diff
@@ -232,6 +232,26 @@ use crate::{fmt, intrinsics, ptr, slice};
 /// remain `#[repr(transparent)]`. That said, `MaybeUninit<T>` will *always* guarantee that it has
 /// the same size, alignment, and ABI as `T`; it's just that the way `MaybeUninit` implements that
 /// guarantee may evolve.
+///
+/// Note that even though `MaybeUninit<T>` and `T` are ABI compatible it is still unsound to
+/// transmute `&mut T` to `&mut MaybeUninit<T>` because that would enable safe Rust to access
+/// uninitialized memory:
```

Comment:
The reasoning only tells me why exposing a `&mut MaybeUninit<T>` from an arbitrary `&mut T` is unsafe, not that it's always *unsound* to do that reinterpretation.

Reply:
Yeah, I don't think this is the right place to put this warning. Though I'm not sure where the right place is. Maybe on `MaybeUninit` itself?

Reply:
This is `MaybeUninit` itself?

Reply:
oh, lol. yeah then only the wording should be changed

Reply:
yeah, I wouldn't say "unsound" because unsafe code inherently can perform this sort of rule-bending feat without it immediately unsound, that's the point of unsafe code... but you do have to restore the invariants before you release control to code that isn't "in on it".

Reply:
Thanks for quick review.

According to the [reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html):
> if `unsafe` code can be misused by safe code to exhibit undefined behavior, it is _unsound_.

so I think the rewording is accurate:

```rs
/// Note that even though `MaybeUninit<T>` and `T` are ABI compatible it is still unsound to
/// transmute `&mut T` to `&mut MaybeUninit<T>` and expose that to safe code because it would allow
/// safe code to access uninitialized memory:
```

Does anyone disagree?

To make it easier to get a feel for it, here is how that looks when rendered:

![image](https://github.com/user-attachments/assets/26cb125b-9122-4123-ad1f-17338680a377)

Reply:
Sure, that seems fine.